tpls/desul/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,521 @@
+  return tmp;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
+#pragma acc routine seq
+inline int device_atomic_min_fetch(
+int* const ptr, const int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  int tmp;
+  tmp = atomicMin(const_cast<int *>(ptr), val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+
+#pragma acc routine seq
+inline unsigned int device_atomic_min_fetch(
+unsigned int* const ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  unsigned int tmp;
+  tmp = atomicMin(const_cast<unsigned int *>(ptr), val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+
+#pragma acc routine seq
+inline unsigned long long device_atomic_min_fetch(
+unsigned long long* const ptr, const unsigned long long val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  unsigned long long tmp;
+  tmp = atomicMin(const_cast<unsigned long long *>(ptr), val);
+  tmp = std::min(tmp, val);
+  return tmp;
+}
+#endif
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_and_fetch(
+    T* const ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  T *lptr = const_cast<T *>(ptr);
+#pragma acc atomic capture
+  { *lptr &= val; tmp = *lptr; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_or_fetch(
+    T* const ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  T *lptr = const_cast<T *>(ptr);
+#pragma acc atomic capture
+  { *lptr |= val; tmp = *lptr; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_xor_fetch(
+    T* const ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  T *lptr = const_cast<T *>(ptr);
+#pragma acc atomic capture
+  { *lptr ^= val; tmp = *lptr; }
+  return tmp;
+}
+//</editor-fold>
+
+//<editor-fold desc="device_atomic_{store,load}">
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,void) device_atomic_store(
+    T* const ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+#pragma acc atomic write
+    *ptr = val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,void) device_atomic_store(
+    T* const ptr, const T val, MemoryOrder, MemoryScopeDevice) {
+  device_atomic_store(ptr, val, MemoryOrderRelaxed(), MemoryScopeDevice()); 
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_load(
+    const T* const ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T retval{};
+#pragma acc atomic read
+  retval = *ptr;
+  return retval;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_load(
+    const T* const ptr, MemoryOrder, MemoryScopeDevice) {
+  T retval{};
+  retval = device_atomic_load(ptr, MemoryOrderRelaxed(), MemoryScopeDevice());
+  return retval;
```

Comment:
```suggestion
  return device_atomic_load(ptr, MemoryOrderRelaxed(), MemoryScopeDevice());
```
Also that is a bug waiting to happen.  Are the ACC atomic construct sequentially consistent?

Reply:
OpenACC has a weak memory model, and thus OpenACC atomic construct is NOT sequentially consistent.